cmake/external_rules.cmake

```diff
@@ -90,6 +92,153 @@ function(download_external_sources)
     OUTPUT_FILE ${PROJECT_BINARY_DIR}/external/output_cmake_build.txt
     WORKING_DIRECTORY ${PROJECT_BINARY_DIR}/external
   )
+
+  if(NOT ANDROID AND NOT IOS)
+    # CMake's find_package(OpenSSL) doesn't quite work right with BoringSSL
+    # unless the header file contains OPENSSL_VERSION_NUMBER.
+    file(READ ${PROJECT_BINARY_DIR}/external/src/boringssl/src/include/openssl/opensslv.h TMP_HEADER_CONTENTS)
+    if (NOT TMP_HEADER_CONTENTS MATCHES OPENSSL_VERSION_NUMBER)
+      file(APPEND ${PROJECT_BINARY_DIR}/external/src/boringssl/src/include/openssl/opensslv.h
+      "\n#ifndef OPENSSL_VERSION_NUMBER\n# define OPENSSL_VERSION_NUMBER  0x10010107L\n#endif\n")
+    endif()
+    # Also add an #include <stdlib.h> since openssl has it and boringssl
+    # doesn't, and some of our code depends on the transitive dependency (this
+    # is a bug).
+    file(READ ${PROJECT_BINARY_DIR}/external/src/boringssl/src/include/openssl/rand.h TMP_HEADER2_CONTENTS)
+    if (NOT TMP_HEADER2_CONTENTS MATCHES "<stdlib.h>")
+      file(APPEND ${PROJECT_BINARY_DIR}/external/src/boringssl/src/include/openssl/rand.h
+      "\n#include <stdlib.h>\n")
+    endif()
+  endif()
+endfunction()
+
+# Builds a subset of external dependencies that need to be built before everything else.
+function(build_external_dependencies)
+  # Setup cmake environment.
+  # These commands are executed from within the currect context, which has set
+  # variables for the target platform. We use "env -i" to clear these
+  # variables, and manually keep the PATH to regular bash path.
+  if(CMAKE_HOST_SYSTEM_NAME STREQUAL "Windows")
+    # Windows doesn't have an 'env' command
+    set(ENV_COMMAND "")
+  else()
+    set(firebase_command_line_path "$ENV{PATH}")
+    set(firebase_command_line_home "$ENV{HOME}")
+    set(ENV_COMMAND env -i PATH=${firebase_command_line_path} HOME=${firebase_command_line_home} )
+  endif()
+
+  message("CMake generator: ${CMAKE_GENERATOR}")
```

Comment:
Recommend using STATUS to ensure these go to stdout. Messages without a mode go to stderr.